<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Making Sense of Metagenomes: Multi-omics integration with DIABLO</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/abacbs/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicons/abacbs/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicons/abacbs/favicon-16x16.png"><link rel="manifest" href="favicons/abacbs/site.webmanifest"><link rel="mask-icon" href="favicons/abacbs/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav abacbs"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="abacbs-logo" alt="Material for the &amp;quot;Making Sense of Metagenomes&amp;quot; workshop for ABACBS 2025 in Adelaide, SA." src="assets/images/abacbs-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/03-multi-omics.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav abacbs" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Material for the &amp;quot;Making Sense of Metagenomes&amp;quot; workshop for ABACBS 2025 in Adelaide, SA." src="assets/images/abacbs-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Making Sense of Metagenomes
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Making Sense of Metagenomes
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Making Sense of Metagenomes
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 40%" class="percentage">
    40%
  </div>
  <div class="progress abacbs">
    <div class="progress-bar abacbs" role="progressbar" style="width: 40%" aria-valuenow="40" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/03-multi-omics.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-metagenomic-workflows.html">1. Introduction to metagenomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-taxsea.html">2. Taxon-set enrichment analysis with TaxSEA</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        3. Multi-omics integration with DIABLO
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#setting-up">Setting up</a></li>
<li><a href="#microbiome-data-pre-processing">Microbiome data pre-processing</a></li>
<li><a href="#metabolome-data-pre-processing">Metabolome data pre-processing</a></li>
<li><a href="#initial-analysis-unsupervised-integration-using-pls">Initial analysis: Unsupervised integration using PLS</a></li>
<li><a href="#diablo-integration">DIABLO integration</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="02-taxsea.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="02-taxsea.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Taxon-set enrichment
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Multi-omics integration with DIABLO</h1>
        <p>Last updated on 2025-11-27 |

        <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/edit/main/episodes/03-multi-omics.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we integrate microbiome and metabolomics data to identify
shared biological signals?</li>
<li>What does DIABLO do, and how is it different from single-omics
analyses?</li>
<li>How do we tune a multi-omics model and choose the number of
components and selected features?</li>
<li>How do we interpret DIABLO outputs to understand discriminative
multi-omics signatures?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Preprocess microbiome and metabolomics data for multi-omics
integration (filtering, transformation, normalisation).</li>
<li>Perform unsupervised integration (sPLS) to explore cross-omics
correlations.</li>
<li>Train, tune, and validate a DIABLO model for classification.</li>
<li>Interpret sample plots, feature correlations, loadings, and
multi-omics signatures produced by DIABLO.</li>
</ul></div>
</div>
</div>
</div>
</div>
<p><strong>Author:</strong> Geraldine Kong</p>
<p>This workshop introduces multi-omics integration using DIABLO from
mixOmics R package, adapted from mixOmics’ tutorial <a href="https://mixomics.org/mixdiablo/diablo-tcga-case-study/" class="external-link uri">https://mixomics.org/mixdiablo/diablo-tcga-case-study/</a>.</p>
<p>We will use the dataset from Franzosa et al. (2019) to integrate
microbiome species profiles from shotgun metagenomics with metabolomics
data. The goal is to identify a multi-omics signature that discriminates
healthy controls from patients with <em>Clostridium difficile</em> (CD)
by finding correlated and predictive features across the omics. You will
learn how to tune the model, explore the relationships between features,
and interpret sample groupings.</p>
<section><h2 class="section-heading" id="setting-up">Setting up<a class="anchor" aria-label="anchor" href="#setting-up"></a></h2>
<hr class="half-width"><p>First, load the required packages and the helper functions for data
pre-processing.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">mixOmics</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: MASS</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: lattice</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Loading required package: ggplot2</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
Loaded mixOmics 6.34.0
Thank you for using mixOmics!
Tutorials: http://mixomics.org
Bookdown vignette: https://mixomicsteam.github.io/Bookdown
Questions, issues: Follow the prompts at http://mixomics.org/contact-us
Cite us:  citation('mixOmics')</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Prevalence filter function</span></span>
<span><span class="co">## Usage example: filtered &lt;- filter_prevalence(count_mat, min_prev = 0.20) # To filter features present in 20% of sample</span></span>
<span><span class="co">## Samples as rows, Features as columns</span></span>
<span><span class="va">filter_prev</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">min_prev</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">is.numeric</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">stop</span><span class="op">(</span><span class="st">"Input data must be numeric (samples x features)."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Calculate prevalence (number of non-zero values) per feature</span></span>
<span>  <span class="va">prev</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="va">n_samples</span> <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Convert proportion threshold to number of samples if needed</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">min_prev</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="fu">ceiling</span><span class="op">(</span><span class="va">min_prev</span> <span class="op">*</span> <span class="va">n_samples</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">cutoff</span> <span class="op">&lt;-</span> <span class="va">min_prev</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># Keep only features meeting prevalence requirement</span></span>
<span>  <span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">prev</span> <span class="op">&gt;=</span> <span class="va">cutoff</span></span>
<span>  </span>
<span>  <span class="co"># Display message</span></span>
<span>  <span class="fu">message</span><span class="op">(</span><span class="st">"Kept "</span>, <span class="fu">sum</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span>, <span class="st">" / "</span>, <span class="fu">ncol</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, </span>
<span>          <span class="st">" features ("</span>, <span class="fu">round</span><span class="op">(</span><span class="fu">sum</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span> <span class="op">/</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"%) "</span>,</span>
<span>          <span class="st">"that are present in ≥ "</span>, <span class="va">min_prev</span> <span class="op">*</span> <span class="fl">100</span>, <span class="st">"% of all samples."</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>, <span class="va">keep</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Abundance filter function</span></span>
<span><span class="co">### threshold: proportion cutoff (0.001 = 0.1%)</span></span>
<span><span class="co">### Samples as rows, Features as columns</span></span>
<span><span class="va">filter_microb_abund</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">threshold</span> <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">is.numeric</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">stop</span><span class="op">(</span><span class="st">"Input data must be numeric (samples x features)."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="co"># total count across all samples</span></span>
<span>  <span class="va">total_counts</span> <span class="op">&lt;-</span> <span class="fu">sum</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># total abundance per feature (column sums)</span></span>
<span>  <span class="va">feature_totals</span> <span class="op">&lt;-</span> <span class="fu">colSums</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># proportion of total for each feature</span></span>
<span>  <span class="va">feature_prop</span> <span class="op">&lt;-</span> <span class="va">feature_totals</span> <span class="op">/</span> <span class="va">total_counts</span></span>
<span>  </span>
<span>  <span class="co"># keep features ≥ threshold</span></span>
<span>  <span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">feature_prop</span> <span class="op">&gt;=</span> <span class="va">threshold</span></span>
<span>  </span>
<span>  <span class="co"># Display message</span></span>
<span>  <span class="fu">message</span><span class="op">(</span><span class="st">"Kept "</span>, <span class="fu">sum</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span>, <span class="st">" / "</span>, <span class="fu">ncol</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, </span>
<span>          <span class="st">" features ("</span>, <span class="fu">round</span><span class="op">(</span><span class="fu">sum</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span> <span class="op">/</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"%) "</span>,</span>
<span>          <span class="st">"with ≥ "</span>, <span class="va">threshold</span> <span class="op">*</span> <span class="fl">100</span>, <span class="st">"% total abundance"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">x</span><span class="op">[</span>, <span class="va">keep</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co"># rCLR normalization function</span></span>
<span><span class="co">### Samples as rows, Features as columns</span></span>
<span><span class="va">rclr_norm</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">pseudocount</span> <span class="op">=</span> <span class="fl">1</span>, <span class="va">base</span> <span class="op">=</span> <span class="fu">exp</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">is.numeric</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">stop</span><span class="op">(</span><span class="st">"Input data must be numeric (samples x features)."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="va">x</span> <span class="op">+</span> <span class="va">pseudocount</span> <span class="co"># add pseudocount to avoid log(0)</span></span>
<span>  </span>
<span>  <span class="va">lx</span> <span class="op">&lt;-</span> <span class="fu">log</span><span class="op">(</span><span class="va">x</span>, base <span class="op">=</span> <span class="va">base</span><span class="op">)</span> <span class="co"># log-transform</span></span>
<span>  </span>
<span>  <span class="co"># robust center per sample (median of log-values)</span></span>
<span>  <span class="va">centers</span> <span class="op">&lt;-</span> <span class="fu">matrixStats</span><span class="fu">::</span><span class="fu">rowMedians</span><span class="op">(</span><span class="va">lx</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># subtract median from each sample</span></span>
<span>  <span class="va">lx_centered</span> <span class="op">&lt;-</span> <span class="fu">sweep</span><span class="op">(</span><span class="va">lx</span>, <span class="fl">1</span>, <span class="va">centers</span>, FUN <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">lx_centered</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Filter and transformation function for metabolomics data</span></span>
<span><span class="co">### Samples as rows, features as columns</span></span>
<span><span class="va">filter_metab_abund</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">cutoff</span> <span class="op">=</span> <span class="fl">1000</span>, <span class="va">min_prop</span> <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Data: samples x features numeric matrix / data.frame</span></span>
<span>  <span class="co"># cutoff: intensity threshold</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">is.numeric</span><span class="op">(</span><span class="fu">as.matrix</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">stop</span><span class="op">(</span><span class="st">"Input x must be numeric (samples x features)."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu">apply</span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">mean</span><span class="op">(</span><span class="va">x</span> <span class="op">&gt;</span> <span class="va">cutoff</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="va">min_prop</span><span class="op">)</span></span>
<span>  <span class="va">filtered</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span>, <span class="va">keep</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="fu">attr</span><span class="op">(</span><span class="va">filtered</span>, <span class="st">"removed_features"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span><span class="op">!</span><span class="va">keep</span><span class="op">]</span></span>
<span>  <span class="fu">attr</span><span class="op">(</span><span class="va">filtered</span>, <span class="st">"kept_features"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">filtered</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Display message</span></span>
<span>  <span class="fu">message</span><span class="op">(</span><span class="st">"Kept "</span>, <span class="fu">sum</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span>, <span class="st">" / "</span>, <span class="fu">ncol</span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, </span>
<span>          <span class="st">" features ("</span>, <span class="fu">round</span><span class="op">(</span><span class="fu">sum</span><span class="op">(</span><span class="va">keep</span><span class="op">)</span> <span class="op">/</span> <span class="fu">ncol</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"%) "</span>,</span>
<span>          <span class="st">"that passed the threshold in ≥ "</span>, <span class="va">min_prop</span> <span class="op">*</span> <span class="fl">100</span>, <span class="st">"% of samples."</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">filtered</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Transform metabolomics data</span></span>
<span><span class="va">transform_metab</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">pseudocount</span>    <span class="op">=</span> <span class="fl">1e-6</span>, <span class="co"># constant added constant to avoid log(0)</span></span>
<span>                            <span class="va">center</span> <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">scale</span>  <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># metab: samples x features numeric matrix / data.frame</span></span>
<span>  </span>
<span>  <span class="co"># coerce to matrix and check</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">is.numeric</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">stop</span><span class="op">(</span><span class="st">"Input 'metab' must be numeric (samples x features)."</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>  <span class="va">rn</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## 1. Sample-wise median normalization (PQN-like)</span></span>
<span>  <span class="va">metab_norm</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">apply</span><span class="op">(</span><span class="va">x</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span> <span class="op">/</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu">median</span><span class="op">(</span><span class="va">x</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## 2. Log10 transform</span></span>
<span>  <span class="va">metab_log</span> <span class="op">&lt;-</span> <span class="fu">log10</span><span class="op">(</span><span class="va">metab_norm</span> <span class="op">+</span> <span class="va">pseudocount</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co">## 3. Autoscaling (mean-center + unit variance by default)</span></span>
<span>  <span class="va">metab_scaled</span> <span class="op">&lt;-</span> <span class="fu">scale</span><span class="op">(</span><span class="va">metab_log</span>, center <span class="op">=</span> <span class="va">center</span>, scale <span class="op">=</span> <span class="va">scale</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># restore dimnames</span></span>
<span>  <span class="fu">rownames</span><span class="op">(</span><span class="va">metab_scaled</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">rn</span></span>
<span>  <span class="fu">colnames</span><span class="op">(</span><span class="va">metab_scaled</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cn</span></span>
<span>  </span>
<span>  <span class="kw">return</span><span class="op">(</span><span class="va">metab_scaled</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Next, load the Franzosa2019 dataset and inspect the R objects. The
data has 144 samples: 56 Controls and 88 CD cases, stored in an R list
object containing the following:</p>
<ul><li><p>metadata: 144 rows (samples) with 7 columns (sample
metadata)</p></li>
<li><p>microbiome: 144 rows (samples) with 55882 columns (bacterial
species)</p></li>
<li><p>metabolome: 144 rows (samples) with 367 columns
(metabolites)</p></li>
</ul><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_int</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">'data/Franzosa_IBD_2019.RDS'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Inspect</span></span>
<span><span class="fu">lapply</span><span class="op">(</span><span class="va">data_int</span>, <span class="va">dim</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$metadata
[1] 144   7

$microbiome
[1]   144 55882

$metabolome
[1] 144 367</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## You can view each object in Rstudio using View(data_int$metadata)</span></span>
<span></span>
<span><span class="co"># Save to each object name</span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="va">data_int</span><span class="op">$</span><span class="va">metadata</span></span>
<span><span class="va">metab</span> <span class="op">&lt;-</span> <span class="va">data_int</span><span class="op">$</span><span class="va">metabolome</span></span>
<span><span class="va">microb</span> <span class="op">&lt;-</span> <span class="va">data_int</span><span class="op">$</span><span class="va">microbiome</span></span>
<span></span>
<span><span class="co"># Number of samples in each group</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
     CD Control
     88      56 </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check that the samples are in the same order across all datasets</span></span>
<span><span class="fu">table</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">microb</span><span class="op">)</span> <span class="op">==</span> <span class="va">meta</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span></span>
<span><span class="cn">TRUE</span></span>
<span> <span class="fl">144</span> </span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">metab</span><span class="op">)</span> <span class="op">==</span> <span class="va">meta</span><span class="op">$</span><span class="va">Sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span></span>
<span><span class="cn">TRUE</span></span>
<span> <span class="fl">144</span> </span></code></pre>
</div>
</section><section><h2 class="section-heading" id="microbiome-data-pre-processing">Microbiome data pre-processing<a class="anchor" aria-label="anchor" href="#microbiome-data-pre-processing"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="filter-the-data">Filter the data<a class="anchor" aria-label="anchor" href="#filter-the-data"></a></h3>
<p>Features that are rare or with low abundance should be filtered out
prior to analysis to remove noise from the data. Here we are filtering
out microbiome features that are NOT present in more than 50% of the
samples, AND have less than 0.1% overall abundance.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter based on prevalence </span></span>
<span><span class="co">## Drop rare features that are not present in more than 50% of the samples</span></span>
<span><span class="va">microb_filt</span> <span class="op">&lt;-</span> <span class="fu">filter_prev</span><span class="op">(</span><span class="va">microb</span>, min_prev <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="co"># 50% of samples, or 70%</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Kept 5664 / 55882 features (10.14%) that are present in ≥ 50% of all samples.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter based on relative abundance</span></span>
<span><span class="co">## Drop features that have total of &lt; 0.1%</span></span>
<span><span class="va">microb_filt_0.1</span> <span class="op">&lt;-</span> <span class="fu">filter_microb_abund</span><span class="op">(</span><span class="va">microb_filt</span>, threshold <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Kept 188 / 5664 features (3.32%) with ≥ 0.1% total abundance</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="transform-the-data">Transform the data<a class="anchor" aria-label="anchor" href="#transform-the-data"></a></h3>
<p>Data should be normalised and transformed prior to integration. Here
we apply the robust centered log-ratio transformation (rCLR) method for
microbiome data which is less sensitive to extreme values than the
standard CLR. A small pseudocount is added prior to rCLR transformation
to avoid log(0).</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rCLR transform data</span></span>
<span><span class="va">microb_rclr</span> <span class="op">&lt;-</span> <span class="fu">rclr_norm</span><span class="op">(</span><span class="va">microb_filt_0.1</span>, pseudocount <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 144 188</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="cleanup-names">Cleanup names<a class="anchor" aria-label="anchor" href="#cleanup-names"></a></h3>
<p>Because the original microbiome features contain the full taxonomic
lineage (very long names), the code below shortens them to species names
only.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Fusicatenibacter;s__Fusicatenibacter saccharivorans"
[2] "d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila_A"
[3] "d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;g__Parabacteroides;s__Parabacteroides goldsteinii"
[4] "d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia_A;s__Blautia_A faecis"
[5] "d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes;s__Anaerostipes hadrus_A"
[6] "d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger;s__Gemmiger qucibialis"                       </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Shorten species names</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu">strsplit</span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">';'</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"s__"</span>, <span class="st">""</span>, <span class="fu">colnames</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># New names</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="fu">colnames</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Fusicatenibacter;s__Fusicatenibacter saccharivorans
                                                                                                        "Fusicatenibacter saccharivorans"
d__Bacteria;p__Verrucomicrobiota;c__Verrucomicrobiae;o__Verrucomicrobiales;f__Akkermansiaceae;g__Akkermansia;s__Akkermansia muciniphila_A
                                                                                                              "Akkermansia muciniphila_A"
          d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__Bacteroidales;f__Tannerellaceae;g__Parabacteroides;s__Parabacteroides goldsteinii
                                                                                                            "Parabacteroides goldsteinii"
                          d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Blautia_A;s__Blautia_A faecis
                                                                                                                       "Blautia_A faecis"
                  d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Lachnospirales;f__Lachnospiraceae;g__Anaerostipes;s__Anaerostipes hadrus_A
                                                                                                                  "Anaerostipes hadrus_A"
                       d__Bacteria;p__Firmicutes_A;c__Clostridia;o__Oscillospirales;f__Ruminococcaceae;g__Gemmiger;s__Gemmiger qucibialis
                                                                                                                    "Gemmiger qucibialis" </code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="metabolome-data-pre-processing">Metabolome data pre-processing<a class="anchor" aria-label="anchor" href="#metabolome-data-pre-processing"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="filter-the-data-1">Filter the data<a class="anchor" aria-label="anchor" href="#filter-the-data-1"></a></h3>
<p>Similar to above, metabolome features that are rare should be
removed. Here, we are filtering out metabolites which are NOT present in
more than 50% of the samples.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metab_filt</span> <span class="op">&lt;-</span> <span class="fu">filter_metab_abund</span><span class="op">(</span>x <span class="op">=</span> <span class="va">metab</span>, cutoff <span class="op">=</span> <span class="fl">100</span>, min_prop <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Kept 217 / 367 features (59.13%) that passed the threshold in ≥ 50% of samples.</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="transform-the-data-1">Transform the data<a class="anchor" aria-label="anchor" href="#transform-the-data-1"></a></h3>
<p>Metabolome data here is transformed using.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metab_log</span> <span class="op">&lt;-</span> <span class="fu">transform_metab</span><span class="op">(</span><span class="va">metab_filt</span>, pseudocount <span class="op">=</span> <span class="fl">1e-6</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">metab_log</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 144 217</code></pre>
</div>
</div>
</section><section><h2 class="section-heading" id="initial-analysis-unsupervised-integration-using-pls">Initial analysis: Unsupervised integration using PLS<a class="anchor" aria-label="anchor" href="#initial-analysis-unsupervised-integration-using-pls"></a></h2>
<hr class="half-width"><div id="callout1" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div class="callout-inner">
<div class="callout-content">
<p>Before training DIABLO, it is important to explore the datasets
separately (e.g., PCA, sPLS-DA) and perform unsupervised integration
(e.g., rCCA or sPLS) to understand correlation structure.</p>
</div>
</div>
</div>
<p>Here we use sPLS and request 25 variables from each dataset
(arbitrary) that maximise covariance between microbiome and
metabolome.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select the number of features to keep in the model</span></span>
<span><span class="va">list.keepX</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">25</span><span class="op">)</span> </span>
<span><span class="va">list.keepY</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">25</span>, <span class="fl">25</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># run sPLS</span></span>
<span><span class="va">spls_res</span> <span class="op">&lt;-</span> <span class="fu">spls</span><span class="op">(</span><span class="va">microb_rclr</span>, <span class="va">metab_log</span>, </span>
<span>             keepX <span class="op">=</span> <span class="va">list.keepX</span>, keepY <span class="op">=</span> <span class="va">list.keepY</span>,</span>
<span>             scale <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> </span></code></pre>
</div>
<p>The <strong>correlation circle plot</strong> below is a good starting
point to visualise the correlations between the selected features for
each component, here showing features with |correlation| above 0.5.
Points that cluster together are positively associated with each other,
whereas points that are on the opposite end are negatively associated
with each other.</p>
<p>There is a cluster of microbiome features that are positively
correlated with metabolome features on the first component (purple and
green dots on the right side of the plot), with a small selection of
metabolome with opposite correlations (green dots on the left side of
the plot). There is also a cluster of microbiome features that have a
contribution towards Component 2, with only 1 (out of the 25) top
metabolome features that have some sort of correlative relationship
between the two.</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot features of first PLS</span></span>
<span><span class="fu">plotVar</span><span class="op">(</span><span class="va">spls_res</span>, cutoff <span class="op">=</span> <span class="fl">0.5</span>, title <span class="op">=</span> <span class="st">"(a) microbiome vs metabolomics"</span>, </span>
<span>        legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"microbiome"</span>, <span class="st">"metabolomics"</span><span class="op">)</span>, </span>
<span>        var.names <span class="op">=</span> <span class="cn">FALSE</span>, style <span class="op">=</span> <span class="st">'graphics'</span>, </span>
<span>        pch <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">16</span>, <span class="fl">17</span><span class="op">)</span>, cex <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">'darkorchid'</span>, <span class="st">'lightgreen'</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-spls_corr_circle_plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Below shows the correlation between the top 25 microbiome and
metabolome features, with the rows corresponding to microbiome Component
1 and 2, and columns corresponding to metabolomics Component 1 and 2.
The first component from the microbiome data shows a very strong
correlation (~0.9) with the first component from the metabolomics data,
while the other components show much weaker correlations. This suggests
that most of the shared signal between the two datasets is captured by
the first component.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate correlation of microbiome and metabolomics</span></span>
<span><span class="va">cors</span> <span class="op">&lt;-</span> <span class="fu">cor</span><span class="op">(</span><span class="va">spls_res</span><span class="op">$</span><span class="va">variates</span><span class="op">$</span><span class="va">X</span>, <span class="va">spls_res</span><span class="op">$</span><span class="va">variates</span><span class="op">$</span><span class="va">Y</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">cors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"microb_comp"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">cors</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">cors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">paste0</span><span class="op">(</span><span class="st">"metab_comp"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu">ncol</span><span class="op">(</span><span class="va">cors</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cors</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>             metab_comp1   metab_comp2
microb_comp1  0.89514137 -2.100475e-17
microb_comp2  0.02162159  7.082259e-01</code></pre>
</div>
</section><section><h2 class="section-heading" id="diablo-integration">DIABLO integration<a class="anchor" aria-label="anchor" href="#diablo-integration"></a></h2>
<hr class="half-width"><div class="section level3">
<h3 id="setup-split-train-and-test-data">Setup: Split train and test data<a class="anchor" aria-label="anchor" href="#setup-split-train-and-test-data"></a></h3>
<p>To evaluate the ability of DIABLO to generalise the signatures to new
datasets, we are using 80% of data (115 samples) to train the DIABLO
model to learn discriminative features, and assessing the performance of
the model on the remaining 20% of the data (29 samples).</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">set.seed</span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>   <span class="co"># reproducible</span></span>
<span><span class="va">index</span> <span class="op">&lt;-</span> <span class="fu">sample</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu">nrow</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.8</span> <span class="op">*</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">microb_rclr</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Separate training / testing data</span></span>
<span><span class="co">## Microbiome</span></span>
<span><span class="va">train_microb</span> <span class="op">&lt;-</span> <span class="va">microb_rclr</span><span class="op">[</span><span class="va">index</span>, <span class="op">]</span></span>
<span><span class="va">test_microb</span>  <span class="op">&lt;-</span> <span class="va">microb_rclr</span><span class="op">[</span><span class="op">-</span><span class="va">index</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">## Metabolomics</span></span>
<span><span class="va">train_metab</span> <span class="op">&lt;-</span> <span class="va">metab_log</span><span class="op">[</span><span class="va">index</span>, <span class="op">]</span></span>
<span><span class="va">test_metab</span> <span class="op">&lt;-</span> <span class="va">metab_log</span><span class="op">[</span><span class="op">-</span><span class="va">index</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">## Metadata</span></span>
<span><span class="va">train_meta</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">[</span><span class="va">index</span>,<span class="op">]</span></span>
<span><span class="va">test_meta</span> <span class="op">&lt;-</span> <span class="va">meta</span><span class="op">[</span><span class="op">-</span><span class="va">index</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># Put data in list</span></span>
<span><span class="va">data_diablo</span> <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>microb <span class="op">=</span> <span class="va">train_microb</span>,</span>
<span>                   metab <span class="op">=</span> <span class="va">train_metab</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="setup-design-for-diablo">Setup: Design for DIABLO<a class="anchor" aria-label="anchor" href="#setup-design-for-diablo"></a></h3>
<p>DIABLO requires a design matrix (0-1 scale), which represents the
strength of the relationship to be modeled between two given dataframes.
Values close to 1 prioritise maximising correlations between the
datasets; values close to 0 prioritise the discriminative ability of the
model. Since predictive ability of the model is desired here, we use a
design matrix of 0.1.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Setup design</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fl">0.1</span>, ncol <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">data_diablo</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">data_diablo</span><span class="op">)</span>, </span>
<span>                dimnames <span class="op">=</span> <span class="fu">list</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">data_diablo</span><span class="op">)</span>, <span class="fu">names</span><span class="op">(</span><span class="va">data_diablo</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">diag</span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">design</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>       microb metab
microb    0.0   0.1
metab     0.1   0.0</code></pre>
</div>
<p>With a design in place, the initial DIABLO model can be fitted using
an arbitrary number of components (ncomp = 5) before tuning.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># form basic DIABLO model</span></span>
<span><span class="va">basic.diablo.model</span> <span class="op">=</span> <span class="fu">block.splsda</span><span class="op">(</span>X <span class="op">=</span> <span class="va">data_diablo</span>, Y <span class="op">=</span> <span class="va">train_meta</span><span class="op">$</span><span class="va">Group</span>, ncomp <span class="op">=</span> <span class="fl">5</span>, design <span class="op">=</span> <span class="va">design</span><span class="op">)</span> </span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Design matrix has changed to include Y; each block will be
            linked to Y.</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="tuning-parameters">Tuning parameters<a class="anchor" aria-label="anchor" href="#tuning-parameters"></a></h3>
<div class="section level4">
<h4 id="number-of-components">Number of components<a class="anchor" aria-label="anchor" href="#number-of-components"></a></h4>
<p>To choose the number of components for the final DIABLO model, the
function <code>perf()</code> is run with 10-fold cross-validation
repeated 10 times, computing performance for each omics block
separately.</p>
<p><em>In the case where you have small number of samples (n &lt;
~10/group), you can use the ‘LOO’ (Leave-one-out) cross-validation
method.</em></p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Tune number of components</span></span>
<span><span class="va">perf.diablo</span> <span class="op">=</span> <span class="fu">perf</span><span class="op">(</span><span class="va">basic.diablo.model</span>, validation <span class="op">=</span> <span class="st">'Mfold'</span>, </span>
<span>                   folds <span class="op">=</span> <span class="fl">10</span>, nrepeat <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">perf.diablo</span><span class="op">$</span><span class="va">error.rate</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$microb
       max.dist centroids.dist mahalanobis.dist
comp1 0.1156522      0.1226087        0.1226087
comp2 0.1234783      0.1200000        0.1243478
comp3 0.1339130      0.1200000        0.1200000
comp4 0.1478261      0.1234783        0.1478261
comp5 0.1513043      0.1234783        0.1478261

$metab
       max.dist centroids.dist mahalanobis.dist
comp1 0.1713043      0.1930435        0.1930435
comp2 0.1486957      0.1547826        0.1539130
comp3 0.1269565      0.1373913        0.1313043
comp4 0.1321739      0.1469565        0.1304348
comp5 0.1278261      0.1469565        0.1339130</code></pre>
</div>
<p>The above shows the cross-validated error rates for the microbiome
and metabolomics blocks across 1–5 components, evaluated under three
different prediction distance metrics (max.dist, centroids.dist,
mahalanobis.dist). Lower values indicate better classification
performance.</p>
<p>We focus on the <code>centroids.dist</code> as it is a metric that is
stable with high-dimensional omics and robust when classes are
imbalanced, so it typically gives the most reliable and interpretable
performance estimates.</p>
<ul><li><p>For the microbiome block: The error rate reduces slightly moving
from Component 1 to Component 2. Increasing the number of components
(Components 3–5) give only marginal further reductions.</p></li>
<li><p>For the metabolome block: Component 1 performs weakest.
Increasing to Component 2 improve performance, with the best at
Component 3. After Component 3, error does not continue decreasing —
Components 4–5 fluctuate slightly, meaning no reliable gain.</p></li>
</ul><p>The performance plateaus after Component 2–3, and later components
provide negligible improvement.</p>
<p>You can visualize the error rates across the datasets as below.</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot</span><span class="op">(</span><span class="va">perf.diablo</span><span class="op">)</span> <span class="co"># plot output of tuning</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-plot_tuning_comp-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p><strong>Centroids distance:</strong> First, for each of the classes,
the centroid is calculated using all the training samples associated
with that class. The Euclidean distance of the test sample to the
centroid of training samples are calculated. The class (outcome) will be
assigned to that sample based on whichever class centroid is closest to
that sample. Classifications made using this metric are less susceptible
to outliers within the training set. This metric is best used when the
classes cluster moderately well - which can be determined by plotting
the samples via the <code>plotIndiv()</code> function.</p>
<p><strong>BER:</strong> Average misclassification rate across all
classes. It is a metric that gives equal weight to each class,
regardless of how many samples are in each class. Perfect prediction
will give a BER of 0, whereas a random chance prediction will have BER
of 0.5 (for 2 classes (or worse for more classes). Higher values
indicates worse performance, whereas lower values indicates better
performance.</p>
<p>In DIABLO tuning, adding components is stopped when additional
components do not reduce BER in a consistent and meaningful way. Hence,
in this case, the tuning results suggest that 1 component is enough to
discriminate between the groups. This can occur when a dataset has
strong discrminative signals.</p>
<p><strong>WeightedVote.error.rate:</strong> Prediction is based on a
weighted vote across components, where early components get more weight
(because they usually capture more discriminative signal). This tends to
work well when signal strength is concentrated in the first
component.</p>
<div id="callout2" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div class="callout-inner">
<div class="callout-content">
<p>Note: For visualisations (e.g, sample plots), a minimum of 2
components is required. In such cases, set <code>ncomp = 2</code> in the
final DIABLO model, but interpret the biology primarily using features
from Component 1.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">perf.diablo</span><span class="op">$</span><span class="va">choice.ncomp</span><span class="op">$</span><span class="va">WeightedVote</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>            max.dist centroids.dist mahalanobis.dist
Overall.ER         1              2                3
Overall.BER        1              2                1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set ncomp = 2 for visualisation</span></span>
<span><span class="va">ncomp</span> <span class="op">=</span> <span class="fl">2</span> <span class="co"># else perf.diablo$choice.ncomp$WeightedVote["Overall.BER", "centroids.dist"] </span></span></code></pre>
</div>
</div>
<div class="section level4">
<h4 id="number-of-features">Number of features<a class="anchor" aria-label="anchor" href="#number-of-features"></a></h4>
<p>DIABLO performs sparse variable selection, retaining only the top
number of features in each block with highest covariance while
maximising discrimination. To find the optimal number of features to
select in each dataset , you can use
<code>tune.block.splsda()</code>.</p>
<p>Below, we set a range of values to test for the function to test
iteratively to find the right number of features with the least error
rate (from 5 to 10, and then incrementally by 2 until 20, and then
incrementally by 5 until 30). You can set different ranges for the
different datasets You can also identify the optimal number of features
to select based on previous exploration of each omics data using
sPLS-DA.</p>
<p>In this step, DIABLO evaluates each candidate number of selected
features using internal cross-validation. For every fold, the model
predicts class membership using the centroid-distance rule
(<code>dist = "centroids.dist"</code>), and the feature subset that
gives the lowest cross-validated error is retained as the optimal
keepX.</p>
<p>Here, this is tuned with 10-fold cross validation, but repeated only
once because of time constraints. <em>In practice, a more thorough
tuning process (with increased <code>nrepeat</code> argument) is
recommended.</em> The features are selected based on the centroids
dists, on the internal Mfold validation.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># *Can take awhile to run</span></span>
<span><span class="co"># set grid of values for each component to test</span></span>
<span><span class="va">test.keepX</span> <span class="op">=</span> <span class="fu">list</span> <span class="op">(</span>microb <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">5</span><span class="op">:</span><span class="fl">9</span>, <span class="fu">seq</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">18</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu">seq</span><span class="op">(</span><span class="fl">20</span>,<span class="fl">30</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                   metab <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">5</span><span class="op">:</span><span class="fl">9</span>, <span class="fu">seq</span><span class="op">(</span><span class="fl">10</span>, <span class="fl">18</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu">seq</span><span class="op">(</span><span class="fl">20</span>,<span class="fl">30</span>,<span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tune.diablo</span> <span class="op">=</span> <span class="fu">tune.block.splsda</span><span class="op">(</span>X <span class="op">=</span> <span class="va">data_diablo</span>, </span>
<span>                                Y <span class="op">=</span> <span class="va">train_meta</span><span class="op">$</span><span class="va">Group</span>, ncomp <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                              test.keepX <span class="op">=</span> <span class="va">test.keepX</span>, design <span class="op">=</span> <span class="va">design</span>,</span>
<span>                              validation <span class="op">=</span> <span class="st">'Mfold'</span>, folds <span class="op">=</span> <span class="fl">10</span>, nrepeat <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                              dist <span class="op">=</span> <span class="st">"centroids.dist"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Design matrix has changed to include Y; each block will be
            linked to Y.</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
You have provided a sequence of keepX of length: 13 for block microb and 13 for block metab.
This results in 169 models being fitted for each component and each nrepeat, this may take some time to run, be patient!</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
You can look into the 'BPPARAM' argument to speed up computation time.</code></pre>
</div>
<p>The final number of optimal features selected are as below, first and
second number for each block corresponding to Component 1 and 2
respectively:</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">list.keepX</span> <span class="op">=</span> <span class="va">tune.diablo</span><span class="op">$</span><span class="va">choice.keepX</span> <span class="co"># set the optimal values of features to retain</span></span>
<span><span class="va">list.keepX</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$microb
[1]  5 10

$metab
[1] 25 30</code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="final-diablo-model">Final DIABLO model<a class="anchor" aria-label="anchor" href="#final-diablo-model"></a></h3>
<p>We then run the final DIABLO model using the parameters from above. A
warning message is expected: DIABLO automatically includes the outcome
in the design so that each block’s component is associated with the
group outcome.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">final.diablo.model</span> <span class="op">=</span> <span class="fu">block.splsda</span><span class="op">(</span>X <span class="op">=</span> <span class="va">data_diablo</span>, </span>
<span>                                  Y <span class="op">=</span> <span class="va">train_meta</span><span class="op">$</span><span class="va">Group</span>, </span>
<span>                                  ncomp <span class="op">=</span> <span class="va">ncomp</span>, </span>
<span>                                  keepX <span class="op">=</span> <span class="va">list.keepX</span>, design <span class="op">=</span> <span class="va">design</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Design matrix has changed to include Y; each block will be
            linked to Y.</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a></h3>
<div class="section level4">
<h4 id="sample-plots">Sample plots<a class="anchor" aria-label="anchor" href="#sample-plots"></a></h4>
<p>The plot below from <code>plotDiablo()</code> is a diagnostic plot
that shows how strongly the components extracted from each block are
correlated, as specified by the design matrix. The <code>ncomp</code>
argument selects which component is shown.</p>
<p>As you can see from the figures below, the first components of the
microbiome and metabolome block are highly correlated to each other
(indicated by the large correlation value shown in the bottom left). The
colours and ellipses related to the sample subtypes indicate the
discriminative power of each component to separate controls from CD.</p>
<ul><li>On Component 1: the group centroids are clearly separated although
the confidence ellipses overlap moderately.</li>
<li>On Component 2: the confidence ellipses overlap substantially,
suggesting little additional discriminative signal beyond the first
component</li>
</ul><div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDiablo</span><span class="op">(</span><span class="va">final.diablo.model</span>, ncomp <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/03-multi-omics-rendered-unnamed-chunk-2-1.png" alt="Component 1" class="figure mx-auto d-block"><figcaption>
Component 1
</figcaption></figure><div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDiablo</span><span class="op">(</span><span class="va">final.diablo.model</span>, ncomp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
<figure style="text-align: center"><img src="fig/03-multi-omics-rendered-unnamed-chunk-3-1.png" alt="Component 2" class="figure mx-auto d-block"><figcaption>
Component 2
</figcaption></figure><p>The sample plot below from <code>plotIndiv()</code> function projects
each sample into the latent space of each block, allowing visual
assessment of clustering patterns. Clustering of the samples can be
better assessed with this plot. Here, we can see that both datasets show
comparable separation of controls vs CD, suggesting that discriminatory
information is present in both omics.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotIndiv</span><span class="op">(</span><span class="va">final.diablo.model</span>, ind.names <span class="op">=</span> <span class="cn">FALSE</span>, legend <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          title <span class="op">=</span> <span class="st">'DIABLO Sample Plots'</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The plot below is an arrow plot from <code>plotArrow()</code>
function, showing per-sample agreement across datasets. Short, aligned
arrows imply high agreement across omics; longer or diverging arrows
indicate disagreement.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotArrow</span><span class="op">(</span><span class="va">final.diablo.model</span>, ind.names <span class="op">=</span> <span class="cn">FALSE</span>, legend <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          title <span class="op">=</span> <span class="st">'DIABLO'</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
<div class="section level4">
<h4 id="variable-plots">Variable plots<a class="anchor" aria-label="anchor" href="#variable-plots"></a></h4>
<p>The correlation circle plot below shows the relationship between the
selected features and their respective components.</p>
<p>On Component 1, a group of microbiome features is positively
associated with a set of metabolite features, while another smaller
subset of metabolites shows opposite correlations.</p>
<p>On Component 2, a small number of metabolome features correlate with
microbiome, however, as shown above, those features are not important in
discriminating between the groups so they can be ignored in the
interpretation.</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotVar</span><span class="op">(</span><span class="va">final.diablo.model</span>, var.names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>        style <span class="op">=</span> <span class="st">'graphics'</span>, legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        pch <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">16</span>, <span class="fl">17</span><span class="op">)</span>, cex <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1.5</span>,<span class="fl">1.5</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">'darkorchid'</span>, <span class="st">'lightgreen'</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-correlation_circle_plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You can visualize the loading weights of each selected variable on
each component and each data set using <code>plotLoadings()</code>.
Colours represent the group in which the feature shows the highest
expression (<code>contrib = 'max'</code>) using the median
(<code>method = 'median'</code>).</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotLoadings</span><span class="op">(</span><span class="va">final.diablo.model</span>, comp <span class="op">=</span> <span class="fl">1</span>, </span>
<span>             contrib <span class="op">=</span> <span class="st">'max'</span>, method <span class="op">=</span> <span class="st">'median'</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-plot_loadings-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The circos plot below visualises pairwise correlations between
selected variables from different omics, represented on the side
quadrants. Here, we are visualizing correlations with at least a
strength of 0.7 (<code>cutoff</code> parameter), depicted by the lines
in the center. The outer most ring indicates the block membership and
group-specific expression levels (Control and CD).</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">circosPlot</span><span class="op">(</span><span class="va">final.diablo.model</span>, cutoff <span class="op">=</span> <span class="fl">0.7</span>, line <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           color.blocks<span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">'darkorchid'</span>, <span class="st">'lightgreen'</span><span class="op">)</span>,</span>
<span>           color.cor <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"chocolate3"</span>,<span class="st">"grey20"</span><span class="op">)</span>, size.labels <span class="op">=</span> <span class="fl">1.5</span>, size.variables <span class="op">=</span> <span class="fl">0.6</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-circos_plot-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The clustered image map (CIM) below is a heatmap of the values of
selected multi-omics signature across samples. By default, Euclidean
distance and Complete linkage methods are used.</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># *Can run into "figure margins too large" error</span></span>
<span><span class="fu">cimDiablo</span><span class="op">(</span><span class="va">final.diablo.model</span>,margins <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">8</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
trimming values to [-3, 3] range for cim visualisation. See 'trim' arg in ?cimDiablo</code></pre>
</div>
<figure><img src="fig/03-multi-omics-rendered-cimDiablo-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></div>
</div>
<div class="section level3">
<h3 id="evaluate-model-performance">Evaluate model performance<a class="anchor" aria-label="anchor" href="#evaluate-model-performance"></a></h3>
<p>To assess the performance of the model, we use the
<code>perf()</code> function to perform 10-fold cross-validation
repeated 10 times. In each cross-validation round, for each block in the
DIABLO model, a predicted class is obtained for every test sample. These
block-specific predictions are then combined to produce a single final
prediction per sample, which is used to compute the different error-rate
metrics.</p>
<p><em><strong>MajorityVote</strong>: Each block contributes one class
vote. The predicted class for a sample is simply the class receiving the
highest number of votes across blocks.</em></p>
<p><em>Example: with three omics blocks and classes A and B, if the
block-level predictions for a sample are A, A, B, the MajorityVote
result for that sample is A.</em></p>
<p><em><strong>WeightedVote</strong>: Each block casts a class vote, but
votes are multiplied by block-specific weights that reflect the
correlation between the block’s latent component and the outcome. The
class whose votes sum to the highest weighted total becomes the
predicted class.</em></p>
<p><em>Example: Using the previous block-level predictions (A, A, B) and
block weights 0.3, 0.2, and 0.7, the weighted votes for class A total
0.3 + 0.2 = 0.5, and the weighted vote for class B is 0.7. Therefore,
the WeightedVote result for that sample is B.</em></p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># *Can take awhile to run</span></span>
<span><span class="co"># run repeated CV performance evaluation</span></span>
<span><span class="va">perf.diablo</span> <span class="op">=</span> <span class="fu">perf</span><span class="op">(</span><span class="va">final.diablo.model</span>, validation <span class="op">=</span> <span class="st">'Mfold'</span>, </span>
<span>                   M <span class="op">=</span> <span class="fl">10</span>, nrepeat <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                   dist <span class="op">=</span> <span class="st">'centroids.dist'</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">perf.diablo</span><span class="op">$</span><span class="va">MajorityVote.error.rate</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$centroids.dist
                comp1     comp2
CD          0.2819444 0.2402778
Control     0.1674419 0.1558140
Overall.ER  0.2391304 0.2086957
Overall.BER 0.2246932 0.1980459</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">perf.diablo</span><span class="op">$</span><span class="va">WeightedPredict.error.rate</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                comp1     comp2
CD          0.1013889 0.1388889
Control     0.1581395 0.1441860
Overall.ER  0.1226087 0.1408696
Overall.BER 0.1297642 0.1415375</code></pre>
</div>
<p>From the results above, it can be seen that the error rate is quite
low across the board, suggesting good classification performance. Let’s
try the model on the test set to see how good it is at classifying novel
samples.</p>
<div class="section level4">
<h4 id="prediction-on-test-set">Prediction on test set<a class="anchor" aria-label="anchor" href="#prediction-on-test-set"></a></h4>
<p>We now test the model on the held-out samples using the
<code>predict()</code> function.</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_test</span> <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>microb <span class="op">=</span> <span class="va">test_microb</span>, </span>
<span>                 metab <span class="op">=</span> <span class="va">test_metab</span><span class="op">)</span></span>
<span></span>
<span><span class="va">predict_diablo</span> <span class="op">=</span> <span class="fu">predict</span><span class="op">(</span><span class="va">final.diablo.model</span>, newdata <span class="op">=</span> <span class="va">data_test</span><span class="op">)</span></span></code></pre>
</div>
<p>To inspect the model’s prediction accuracy, we are using a confusion
matrix.</p>
<p><em>A confusion matrix is a table that compares a model’s predicted
labels with the true labels in a classification problem.</em> It tells
you not just how many predictions were wrong, but which classes were
confused with each other. This is especially important when:</p>
<ul><li>Accuracy alone hides which groups perform poorly<br></li>
<li>Some classes are harder to classify than others<br></li>
<li>There is class imbalance</li>
</ul><p>High diagonal values indicate good agreement between predicted and
true labels.</p>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">confusion.mat</span> <span class="op">=</span> <span class="fu">get.confusion_matrix</span><span class="op">(</span>truth <span class="op">=</span> <span class="va">test_meta</span><span class="op">$</span><span class="va">Group</span>,</span>
<span>                                     predicted <span class="op">=</span> <span class="va">predict_diablo</span><span class="op">$</span><span class="va">WeightedVote</span><span class="op">$</span><span class="va">centroids.dist</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">confusion.mat</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        predicted.as.CD predicted.as.Control
CD                   14                    2
Control               2                   11</code></pre>
</div>
<p>The balanced error rate (BER) remains low, confirming that the DIABLO
signature generalises well to the unseen test samples and performs
similarly across both classes.</p>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">get.BER</span><span class="op">(</span><span class="va">confusion.mat</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 0.1394231</code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>DIABLO integrates multiple omics datasets by identifying correlated
and discriminative features across blocks.</li>
<li>Proper filtering and transformation of each omics dataset are
essential for stable multi-omics models.</li>
<li>Tuning (components and keepX) guides the selection of the most
predictive and biologically relevant features.</li>
<li>DIABLO visualisations (sample plots, arrow plots, circos, CIM)
reveal relationships across omics and help interpret multi-omics
signatures.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="02-taxsea.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="index.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="02-taxsea.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Taxon-set enrichment
        </a>
        <a class="chapter-link float-end" href="index.html" rel="next">
          Home
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/edit/main/episodes/03-multi-omics.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/" class="external-link">Source</a></p>
        <p>
        <a href="citation.html">Cite</a>
        | <a href="mailto:vinicius.salazar%20%5B%20at%20%5D%20unimelb.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors.</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/melbournebioinformatics/uom-varnish/tree/main" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "melbournebioinformatics.github.io/uom-varnish/03-multi-omics.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "metagenomics, microbiome, bioinformatics, multiomics, metabolomics, enrichment analysis",
  "name": "Multi-omics integration with DIABLO",
  "creativeWorkStatus": "active",
  "url": "melbournebioinformatics.github.io/uom-varnish/03-multi-omics.html",
  "identifier": "melbournebioinformatics.github.io/uom-varnish/03-multi-omics.html",
  "dateCreated": "2025-10-16",
  "dateModified": "2025-11-27",
  "datePublished": "2025-11-28"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

