<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Making Sense of Metagenomes: Taxon-set enrichment analysis with TaxSEA</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/abacbs/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/abacbs/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/abacbs/favicon-16x16.png"><link rel="manifest" href="../favicons/abacbs/site.webmanifest"><link rel="mask-icon" href="../favicons/abacbs/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav abacbs"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="abacbs-logo" alt="Material for the &amp;quot;Making Sense of Metagenomes&amp;quot; workshop for ABACBS 2025 in Adelaide, SA." src="../assets/images/abacbs-logo.svg"><span class="badge text-bg-warning">
          <abbr title="This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-triangle" style="border-radius: 5px"></i>
              Alpha
            </a>
            <span class="visually-hidden">This lesson is in the alpha phase, which means that it has been taught once and lesson authors are iterating on feedback.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../02-taxsea.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav abacbs" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Material for the &amp;quot;Making Sense of Metagenomes&amp;quot; workshop for ABACBS 2025 in Adelaide, SA." src="../assets/images/abacbs-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Making Sense of Metagenomes
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Making Sense of Metagenomes
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Making Sense of Metagenomes
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 20%" class="percentage">
    20%
  </div>
  <div class="progress abacbs">
    <div class="progress-bar abacbs" role="progressbar" style="width: 20%" aria-valuenow="20" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../02-taxsea.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-metagenomic-workflows.html">1. Introduction to metagenomics</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Taxon-set enrichment analysis with TaxSEA
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#taxon-set-database">Taxon set database</a></li>
<li><a href="#load-the-package">Load the Package</a></li>
<li><a href="#input">Input</a></li>
<li><a href="#load-and-inspect-test-data">Load and Inspect Test Data</a></li>
<li><a href="#run-taxsea-with-test-data">Run TaxSEA with Test Data</a></li>
<li><a href="#examine-outputs">Examine outputs</a></li>
<li><a href="#output">Output</a></li>
<li><a href="#visualisation-of-taxsea-output">Visualisation of TaxSEA output</a></li>
<li><a href="#custom-databases">Custom databases</a></li>
<li><a href="#perform-enrichment-analysis-using-taxsea-with-custom-databases">Perform enrichment analysis using TaxSEA with custom databases</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-multi-omics.html">3. Multi-omics integration with DIABLO</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-metagenomic-workflows.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-multi-omics.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-metagenomic-workflows.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to
        </a>
        <a class="chapter-link float-end" href="../instructor/03-multi-omics.html" rel="next">
          Next: Multi-omics...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Taxon-set enrichment analysis with TaxSEA</h1>
        <p>Last updated on 2025-11-25 |

        <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/edit/main/episodes/02-taxsea.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 20 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can we test whether groups of related taxa (taxon sets) show
coordinated shifts between conditions?</li>
<li>What input does TaxSEA need, and how should taxonomic IDs be
formatted?</li>
<li>How does TaxSEA extend differential abundance results using
enrichment analysis?</li>
<li>How can we interpret enriched taxon sets to understand functional,
ecological, or disease-associated patterns?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Prepare species- or genus-level differential abundance results for
use with TaxSEA.</li>
<li>Run TaxSEA to test for enrichment across predefined or custom taxon
sets.</li>
<li>Understand the structure of TaxSEA output, including P values,
effect direction, and FDR.</li>
<li>Interpret enrichment results in the context of microbiome function,
ecology, and host associations.</li>
</ul></div>
</div>
</div>
</div>
</div>
<p><strong>Authors:</strong> Calum J. Walsh &amp; Feargal J. Ryan</p>
<p>TaxSEA is a Bioconductor package that helps microbiome researchers
test for enrichment in known microbial signatures, including:</p>
<ul><li>Metabolite producers</li>
<li>Disease associations</li>
<li>Previously published microbiome signatures</li>
<li>Traits established from <em>in-vitro</em> data</li>
</ul><p>TaxSEA takes as input a vector of genus or species names and a rank
(e.g. log2 fold changes or Spearman’s rho) and uses a Kolmogorov-Smirnov
test to identify if a particular group of species or genera (i.e. a set
of taxa such as butyrate producers) are skewed to one end of the
distribution (i.e. enriched).</p>
<p>TaxSEA is based on the concept of Gene Set Enrichment Analysis
(GSEA). For an overview, check out the link here:<a href="https://www.metwarebio.com/gsea-enrichment-analysis-guide/" class="external-link">https://www.metwarebio.com/gsea-enrichment-analysis-guide/</a>.</p>
<p>Note: Although TaxSEA can, in principle, be applied to microbiome
data from any source, the databases utilised largely cover
human-associated microbiomes, particularly the human gut microbiome. As
such TaxSEA will likely perform best on human gut microbiome data.</p>
<section><h2 class="section-heading" id="taxon-set-database">Taxon set database<a class="anchor" aria-label="anchor" href="#taxon-set-database"></a></h2>
<hr class="half-width"><p>By default TaxSEA utilises taxon sets generated from reference
databases (gutMGene, GMrepo v2, MiMeDB, mBodyMap, BugSigDB, BacDive) as
well as a handful of sets curated from the literature (mucin utilisers
and BLOSSUM taxa). See below for examples of using custom databases or
taxonomically-defined taxon sets.</p>
<p>Please cite the appropriate database if using:</p>
<ul><li>Cheng <em>et al.</em> gutMGene: a comprehensive database for target
genes of gut microbes and microbial metabolites Nucleic Acids Res.
2022.</li>
<li>Dai <em>et al.</em> GMrepo v2: a curated human gut microbiome
database with special focus on disease markers and cross-dataset
comparison Nucleic Acids Res. 2022.</li>
<li>Wishart <em>et. al.</em> MiMeDB: the Human Microbial Metabolome
Database Nucleic Acids Res. 2023.</li>
<li>Jin <em>et al.</em> mBodyMap: a curated database for microbes across
human body and their associations with health and diseases. Nucleic
Acids Res. 2022.</li>
<li>Geistlinger <em>et al.</em> BugSigDB captures patterns of
differential abundance across a broad range of host-associated microbial
signatures. Nature Biotech. 2023.</li>
</ul></section><section><h2 class="section-heading" id="load-the-package">Load the Package<a class="anchor" aria-label="anchor" href="#load-the-package"></a></h2>
<hr class="half-width"><div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">TaxSEA</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a></h2>
<hr class="half-width"><p>All that is required for TaxSEA is an R vector containing ranks
(e.g. log2 fold changes) and names (e.g. species/genus). TaxSEA will not
work if input names are from ranks higher than species or genus and will
struggle with GTDB taxonomy. The input should include all taxa tested,
not a limited or pre-defined set (e.g. do not use a threshold for
significance or remove any taxa). TaxSEA will lookup and convert taxon
names to NCBI taxonomic identifiers. TaxSEA stores a commonly observed
identifiers internally and so will only lookup whatever is not covered
to save time.</p>
<p>Input IDs should be formatted like one of the following:</p>
<ul><li>Species name. E.g. “Bifidobacterium longum”,
“Bifidobacterium_longum”</li>
<li>Genus name. E.g. “Bifidobacterium”</li>
<li>NCBI ID E.g. “216816”</li>
</ul><p>The code chunk below is useful for splitting names that contain the
full species taxonomy (eg. MetaPhlAn output). We don’t need to run it
now.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Input IDs with the full taxonomic lineage should be split E.g.</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="st">"d__Bacteria.p__Actinobacteriota.c__Actinomycetes.o__Bifidobacteriales.f__Bifidobacteriaceae.g__Bifidobacterium"</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">strsplit</span><span class="op">(</span><span class="va">x</span>, split <span class="op">=</span> <span class="st">"\\."</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"g__"</span>, <span class="st">""</span>, <span class="va">x</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Running this through a vector of IDs may look something like the following</span></span>
<span><span class="va">new_ids</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="va">old_ids</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">{</span><span class="fu">strsplit</span><span class="op">(</span>x <span class="op">=</span> <span class="va">y</span>, split <span class="op">=</span> <span class="st">"\\."</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">new_ids</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"g__"</span>, <span class="st">""</span>, <span class="va">new_ids</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="load-and-inspect-test-data">Load and Inspect Test Data<a class="anchor" aria-label="anchor" href="#load-and-inspect-test-data"></a></h2>
<hr class="half-width"><p>The TaxSEA library comes with pre-processed test data included.<br>
The data comes from a study <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-017-0490-5" class="external-link">comparing
IBD patients to healthy controls</a>.<br>
The count data was downloaded from <a href="https://www.nature.com/articles/nmeth.4468" class="external-link">curatedMetagenomicData</a>
and fold changes between the two groups were generated with <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02655-5" class="external-link">LinDA</a>.</p>
<p>We can load this data into our working environment using the
<code>data()</code> function</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data</span><span class="op">(</span><span class="st">"TaxSEA_test_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">sample</span><span class="op">(</span><span class="va">TaxSEA_test_data</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     Bifidobacterium_bifidum Faecalibacterium_prausnitzii
                       3.146                       -4.040
           Eggerthella_lenta Butyricicoccus_pullicaecorum
                       1.335                        2.021 </code></pre>
</div>
<p>The input is a vector of species names and log fold changes. Perfect
for TaxSEA.</p>
</section><section><h2 class="section-heading" id="run-taxsea-with-test-data">Run TaxSEA with Test Data<a class="anchor" aria-label="anchor" href="#run-taxsea-with-test-data"></a></h2>
<hr class="half-width"><div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxsea_results</span> <span class="op">&lt;-</span> <span class="fu">TaxSEA</span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">TaxSEA_test_data</span><span class="op">)</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="examine-outputs">Examine outputs<a class="anchor" aria-label="anchor" href="#examine-outputs"></a></h2>
<hr class="half-width"></section><section><h2 class="section-heading" id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a></h2>
<hr class="half-width"><p>The output is a list of data frames providing enrichment results for
metabolite produers, health/disease associations, and published
signatures from BugSigDB. Each dataframe has at least 6 columns:</p>
<ol style="list-style-type: decimal"><li>
<em>taxonSetName</em> - The name of the taxon set tested</li>
<li>
<em>median_rank_of_set_members</em> - This is simply the median rank
across all detected members in the set. This allows you to see the
direction of change</li>
<li>
<em>PValue</em> - Kolmogorov-Smirnov test P value.</li>
<li>
<em>Test_statistic</em> - Kolmogorov-Smirnov test statistic.</li>
<li>
<em>FDR</em> - P value adjusted for multiple testing.</li>
<li>
<em>TaxonSet</em> - Returns list of taxa in the set to show what is
driving the signal</li>
</ol><p><strong>All results:</strong> This includes all databases. Is useful
for a quick look and overview but can get a bit messy.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_results.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">All_databases</span></span></code></pre>
</div>
<p><strong>Metabolite Producers:</strong> Enrichment among metabolite
producers from gutMgene and MiMeDB</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metabolites.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Metabolite_producers</span></span></code></pre>
</div>
<p><strong>Health Associations:</strong> Enrichment among health and
disease signatures from GMRepoV2 and mBodyMap</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">disease.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">Health_associations</span></span></code></pre>
</div>
<p><strong>BugSigDB:</strong> Enrichment among published associations
from BugSigDB - a database of manually curated microbial signatures from
published differential abundance studies. The output format here is a
bit different. You’ll notice that, where available, it includes a PubMed
ID linking back to the reporting paper.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bsdb.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BugSigDB</span></span></code></pre>
</div>
<p><strong>Bacterial</strong> <strong>Physiology:</strong> Enrichment
among bacterial meta data from from BacDive - a standardised database of
bacterial information, including physiology.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bacdive.df</span> <span class="op">&lt;-</span> <span class="va">taxsea_results</span><span class="op">$</span><span class="va">BacDive_bacterial_physiology</span></span></code></pre>
</div>
</section><section><h2 class="section-heading" id="visualisation-of-taxsea-output">Visualisation of TaxSEA output<a class="anchor" aria-label="anchor" href="#visualisation-of-taxsea-output"></a></h2>
<hr class="half-width"><p>As TaxSEA operates on individual species and ranks, the easiest way
to visualise these is often through the input ranks. For example using
volcano plots, or density plots can be intuitive and useful ways to see
the overall trend in a set and the members driving it. Remember just
becomes an individual species isn’t significant by itself, doesn’t mean
that the set cannot be significant.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">reshape</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">curatedMetagenomicData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">TaxSEA</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">MicrobiomeStat</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggrepel</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggpubr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">ggridges</span><span class="op">)</span></span>
<span></span>
<span><span class="co">########################</span></span>
<span><span class="co">#### HallAB_2017 ####</span></span>
<span><span class="co">########################</span></span>
<span><span class="va">allmeta_data</span> <span class="op">=</span> <span class="va">sampleMetadata</span></span>
<span></span>
<span><span class="co"># HallAB_2017 dataset</span></span>
<span><span class="va">HallAB_2017_cmd_object</span> <span class="op">=</span> <span class="fu">curatedMetagenomicData</span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"2021-10-14.HallAB_2017.relative_abundance"</span>,counts <span class="op">=</span> <span class="cn">TRUE</span>,dryrun <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HallAB_2017_counts.df</span> <span class="op">=</span> <span class="va">HallAB_2017_cmd_object</span></span>
<span><span class="va">HallAB_2017_counts.df</span> <span class="op">=</span> <span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">$</span><span class="va">`2021-10-14.HallAB_2017.relative_abundance`</span><span class="op">)</span></span>
<span></span>
<span><span class="va">HallAB_2017_counts.df</span> <span class="op">=</span> <span class="fu">assay</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">)</span></span>
<span><span class="va">HallAB_2017_counts.df</span> <span class="op">=</span> <span class="va">HallAB_2017_counts.df</span><span class="op">[</span><span class="fu">apply</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">&gt;</span><span class="fl">1000</span>,<span class="fl">1</span>,<span class="va">sum</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">4</span>,<span class="op">]</span></span>
<span><span class="va">HallAB_2017_md.df</span> <span class="op">=</span> <span class="va">allmeta_data</span><span class="op">[</span><span class="va">allmeta_data</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">%in%</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_md.df</span><span class="op">)</span> <span class="op">=</span> <span class="va">HallAB_2017_md.df</span><span class="op">$</span><span class="va">sample_id</span></span>
<span></span>
<span><span class="co">## Ensure data is matching</span></span>
<span><span class="fu">colnames</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_md.df</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
 [16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
 [31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
 [46] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
 [61] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
 [76] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
 [91] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[106] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[121] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[136] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[151] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[166] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[181] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[196] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[211] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[226] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[241] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE
[256] TRUE TRUE TRUE TRUE</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HallAB_2017_counts.df</span> <span class="op">=</span> <span class="va">HallAB_2017_counts.df</span><span class="op">[</span>,<span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_md.df</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">HallAB_2017_md.df</span> <span class="op">=</span> <span class="va">HallAB_2017_md.df</span><span class="op">[</span>,<span class="fu">c</span><span class="op">(</span><span class="st">"study_name"</span>,<span class="st">"sample_id"</span>,<span class="st">"subject_id"</span>,<span class="st">"study_condition"</span>,<span class="st">"age_category"</span>,<span class="st">"DNA_extraction_kit"</span>,<span class="st">"visit_number"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Remove duplicate samples </span></span>
<span><span class="va">HallAB_2017_md.df</span> <span class="op">=</span> <span class="va">HallAB_2017_md.df</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">HallAB_2017_md.df</span><span class="op">$</span><span class="va">visit_number</span>,decreasing <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">HallAB_2017_md.df</span> <span class="op">=</span> <span class="va">HallAB_2017_md.df</span><span class="op">[</span><span class="op">!</span><span class="fu">duplicated</span><span class="op">(</span><span class="va">HallAB_2017_md.df</span><span class="op">$</span><span class="va">subject_id</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">HallAB_2017_counts.df</span> <span class="op">=</span> <span class="va">HallAB_2017_counts.df</span><span class="op">[</span>,<span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_md.df</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Assigning rows as species names ##</span></span>
<span><span class="va">spec_vec</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">as.character</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">)</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">{</span><span class="fu">strsplit</span><span class="op">(</span>x <span class="op">=</span> <span class="va">y</span>,split<span class="op">=</span><span class="st">"\\|"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">spec_vec</span><span class="op">)</span> <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="va">spec_vec</span> <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"s__"</span>,<span class="st">""</span>,<span class="va">spec_vec</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span><span class="op">)</span> <span class="op">=</span> <span class="va">spec_vec</span></span>
<span></span>
<span><span class="va">HallAB_2017_linda_res</span> <span class="op">=</span> <span class="fu">linda</span><span class="op">(</span><span class="va">HallAB_2017_counts.df</span>, </span>
<span>                              <span class="va">HallAB_2017_md.df</span>, </span>
<span>                              formula <span class="op">=</span> <span class="st">'~study_condition+DNA_extraction_kit'</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>0  features are filtered!
The filtered data has  32  samples and  299  features will be tested!</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Imputation approach is used.
Fit linear models ...
Completed.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HallAB_2017_ranks</span> <span class="op">=</span> <span class="va">HallAB_2017_linda_res</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">study_conditionIBD</span><span class="op">$</span><span class="va">log2FoldChange</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">HallAB_2017_ranks</span><span class="op">)</span> <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_linda_res</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">study_conditionIBD</span><span class="op">)</span></span>
<span><span class="va">HallAB_2017_TaxSEA_results.df</span> <span class="op">=</span> <span class="fu">TaxSEA</span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">HallAB_2017_ranks</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alldb</span> <span class="op">=</span> <span class="va">HallAB_2017_TaxSEA_results.df</span><span class="op">$</span><span class="va">All_databases</span>  </span>
<span><span class="va">HallAB_2017_DA</span> <span class="op">=</span> <span class="va">HallAB_2017_linda_res</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">study_conditionIBD</span></span>
<span></span>
<span><span class="va">hall_bacterialids</span> <span class="op">=</span> <span class="fu">get_ncbi_taxon_ids</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_DA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">data</span><span class="op">(</span><span class="st">"TaxSEA_db"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scfas</span> <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">TaxSEA_db</span><span class="op">$</span><span class="va">GutMGene_producers_of_Propionate</span>,</span>
<span>                        <span class="va">TaxSEA_db</span><span class="op">$</span><span class="va">GutMGene_producers_of_Butyrate</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">NCBI_id</span> <span class="op">=</span> <span class="va">hall_bacterialids</span><span class="op">[</span><span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_DA</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">Nitrate_utilisers</span> <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">NCBI_id</span> <span class="op">%in%</span> <span class="va">TaxSEA_db</span><span class="op">$</span><span class="va">BacDive_Utilizes_nitrate</span></span>
<span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">Faculatative_anerobes</span> <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">NCBI_id</span> <span class="op">%in%</span> <span class="va">TaxSEA_db</span><span class="op">$</span><span class="va">`BacDive_facultative anaerobe`</span></span>
<span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">SCFA_producers</span> <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">NCBI_id</span> <span class="op">%in%</span> <span class="va">scfas</span></span>
<span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">Species</span> <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">HallAB_2017_DA</span><span class="op">)</span></span>
<span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">mBodyMap_skin</span> <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">NCBI_id</span> <span class="op">%in%</span> <span class="va">TaxSEA_db</span><span class="op">$</span><span class="va">mBodyMap_skin</span></span>
<span></span>
<span></span>
<span><span class="va">volcano_plot_all</span> <span class="op">=</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">HallAB_2017_DA</span>,<span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">log2FoldChange</span>,y<span class="op">=</span><span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span>color<span class="op">=</span><span class="st">"grey33"</span>,size<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu">guides</span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span>,size<span class="op">=</span><span class="st">"none"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_label_repel</span><span class="op">(</span>data<span class="op">=</span><span class="va">HallAB_2017_DA</span><span class="op">[</span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.1</span>,<span class="op">]</span>,</span>
<span>                   <span class="fu">aes</span><span class="op">(</span>label<span class="op">=</span><span class="va">Species</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>,linetype<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>,linetype<span class="op">=</span><span class="fl">3</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"All Species"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">scfa_volcano</span> <span class="op">=</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">HallAB_2017_DA</span>,<span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">log2FoldChange</span>,y<span class="op">=</span><span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>,</span>
<span>                               colour<span class="op">=</span><span class="va">SCFA_producers</span>,</span>
<span>                               size<span class="op">=</span><span class="va">SCFA_producers</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">scale_colour_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"grey33"</span>,<span class="st">"deepskyblue3"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">scale_size_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu">guides</span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span>,size<span class="op">=</span><span class="st">"none"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_label_repel</span><span class="op">(</span>data<span class="op">=</span><span class="va">HallAB_2017_DA</span><span class="op">[</span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">SCFA_producers</span><span class="op">==</span><span class="cn">TRUE</span>,<span class="op">]</span>,</span>
<span>                   <span class="fu">aes</span><span class="op">(</span>label<span class="op">=</span><span class="va">Species</span><span class="op">)</span>,max.overlaps <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"SCFA producers"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">volcano_fa</span> <span class="op">=</span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">HallAB_2017_DA</span>,<span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">log2FoldChange</span>,y<span class="op">=</span><span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">padj</span><span class="op">)</span>,</span>
<span>                               colour<span class="op">=</span><span class="va">Faculatative_anerobes</span>,</span>
<span>                               size<span class="op">=</span><span class="va">Faculatative_anerobes</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">scale_colour_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"grey33"</span>,<span class="st">"darkorchid"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">scale_size_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span><span class="op">+</span><span class="fu">guides</span><span class="op">(</span>color<span class="op">=</span><span class="st">"none"</span>,size<span class="op">=</span><span class="st">"none"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">geom_label_repel</span><span class="op">(</span>data<span class="op">=</span><span class="va">HallAB_2017_DA</span><span class="op">[</span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">Faculatative_anerobes</span><span class="op">==</span><span class="cn">TRUE</span>,<span class="op">]</span>,</span>
<span>                   <span class="fu">aes</span><span class="op">(</span>label<span class="op">=</span><span class="va">Species</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Faculatative anerobes"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ggarrange</span><span class="op">(</span><span class="va">volcano_plot_all</span>,<span class="va">volcano_fa</span>,<span class="va">scfa_volcano</span>,ncol <span class="op">=</span> <span class="fl">3</span>,nrow<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-taxsea-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="co">## Build a long df for the ridge plot</span></span>
<span> <span class="va">ridge_dat</span> <span class="op">&lt;-</span> <span class="fu">rbind</span><span class="op">(</span></span>
<span>   <span class="fu">data.frame</span><span class="op">(</span>log2FoldChange <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">log2FoldChange</span>,</span>
<span>              Group <span class="op">=</span> <span class="st">"All taxa"</span><span class="op">)</span>,</span>
<span>   <span class="fu">data.frame</span><span class="op">(</span>log2FoldChange <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">[</span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">SCFA_producers</span><span class="op">]</span>,</span>
<span>              Group <span class="op">=</span> <span class="st">"SCFA producers"</span><span class="op">)</span>,</span>
<span>   <span class="fu">data.frame</span><span class="op">(</span>log2FoldChange <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">[</span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">mBodyMap_skin</span><span class="op">]</span>,</span>
<span>              Group <span class="op">=</span> <span class="st">"Skin-associated"</span><span class="op">)</span>,</span>
<span>   <span class="fu">data.frame</span><span class="op">(</span>log2FoldChange <span class="op">=</span> <span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">[</span><span class="va">HallAB_2017_DA</span><span class="op">$</span><span class="va">Faculatative_anerobes</span><span class="op">]</span>,</span>
<span>              Group <span class="op">=</span> <span class="st">"Facultative anaerobes"</span><span class="op">)</span></span>
<span> <span class="op">)</span></span>
<span> </span>
<span> <span class="va">ridge_dat</span><span class="op">$</span><span class="va">Group</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span></span>
<span>   <span class="va">ridge_dat</span><span class="op">$</span><span class="va">Group</span>,</span>
<span>   levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"All taxa"</span>,</span>
<span>                  <span class="st">"SCFA producers"</span>,</span>
<span>                  <span class="st">"Skin-associated"</span>,</span>
<span>                  <span class="st">"Facultative anaerobes"</span><span class="op">)</span><span class="op">)</span></span>
<span> <span class="op">)</span></span>
<span> </span>
<span> <span class="fu">ggplot</span><span class="op">(</span><span class="va">ridge_dat</span>,</span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2FoldChange</span>,</span>
<span>            y <span class="op">=</span> <span class="va">Group</span>,</span>
<span>            fill <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">geom_density_ridges</span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.85</span>, scale <span class="op">=</span> <span class="fl">1.2</span>, colour <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">xlim</span><span class="op">(</span><span class="op">-</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span></span>
<span>     <span class="st">"grey44"</span>,        <span class="co"># All taxa</span></span>
<span>     <span class="st">"deepskyblue3"</span>,  <span class="co"># SCFA</span></span>
<span>     <span class="st">"tan3"</span>,          <span class="co"># Skin</span></span>
<span>     <span class="st">"darkorchid"</span>     <span class="co"># Anaerobes</span></span>
<span>   <span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ylab</span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"log2FC distributions for functional/ecological groups"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-taxsea-rendered-unnamed-chunk-10-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="custom-databases">Custom databases<a class="anchor" aria-label="anchor" href="#custom-databases"></a></h2>
<hr class="half-width"><p>If you want to run TaxSEA with your own database, use the
<code>custom_db</code> argument. It should be a named list where:</p>
<ul><li>each <strong>name</strong> is the label for a set, and</li>
<li>each <strong>value</strong> is a vector of species assigned to that
set.</li>
</ul><p>This is the same format as the default TaxSEA database.</p>
<p>Note: using the <em>custom_db</em> flag disables the automatic ID
conversion and NCBI API lookup. However this is available via other
functions.</p>
</section><section><h2 class="section-heading" id="perform-enrichment-analysis-using-taxsea-with-custom-databases">Perform enrichment analysis using TaxSEA with custom databases<a class="anchor" aria-label="anchor" href="#perform-enrichment-analysis-using-taxsea-with-custom-databases"></a></h2>
<hr class="half-width"><p>In addition to taxon sets defined by function or phenotype, users can
define custom sets based on taxonomy. Current methods to test
differential abundance at higher taxonomic levels (e.g., genus or
family) involve aggregating counts, but with this approach opposing
shifts in individual species may cancel each other out, obscuring
meaningful biological patterns. For instance, antibiotic treatment may
suppress certain species while allowing resistant species within the
same genus to expand and occupy the vacant niche, creating an ecological
shift that appears as no net change at broader taxonomic levels. Here we
utilise data from <a href="https://pubmed.ncbi.nlm.nih.gov/27562258/" class="external-link">Chng et al. 2016</a>, a
study comparing the skin microbiome between Atopic dermatitis and
controls.</p>
<p>The script below applies the TaxSEA framework to identify taxonomic
enrichment at different taxonomic levels. Specifically, we analyse
enrichment at the order level from metagenomic data.</p>
<p>The code block below does the following:</p>
<ul><li>Downloads species abundance data</li>
<li>Performs some light QC</li>
<li>Defines taxonomic sets at each level of taxonomy (phylum -&gt;
genus)</li>
<li>Calculates differential abundance between groups</li>
<li>Performs enrichment analysis using TaxSEA</li>
<li>Plots results</li>
</ul><!--
data.frame(Name = names(log2_fold_changes),

Log2FC = log2_fold_changes) %\>%

left_join(taxon_lineages) %\>%

left_join(custom_taxsea_results, by = c("taxonSetName" = "order")) %\>%

drop_na(PValue)
 --><div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required libraries</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">TaxSEA</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">curatedMetagenomicData</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">phyloseq</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">MicrobiomeStat</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DT</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load sample metadata from curatedMetagenomicData</span></span>
<span><span class="va">metadata_all</span> <span class="op">&lt;-</span> <span class="va">sampleMetadata</span></span>
<span></span>
<span><span class="co"># Filter metadata for the specific study (ChngKR_2016)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata_all</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">study_name</span> <span class="op">==</span> <span class="st">"ChngKR_2016"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'sample_id'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract count data from curatedMetagenomicData </span></span>
<span><span class="va">cmd_data</span> <span class="op">&lt;-</span> <span class="fu">curatedMetagenomicData</span><span class="op">(</span></span>
<span>  pattern <span class="op">=</span> <span class="st">"ChngKR_2016.relative_abundance"</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  dryrun <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert the extracted data to a count matrix</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">cmd_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co"># Subset to relevant samples</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">counts_data</span><span class="op">[</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span><span class="op">]</span>  </span>
<span></span>
<span><span class="co"># Filter taxa to only keep those with at least one sample having counts &gt; 100</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="va">counts_data</span><span class="op">[</span><span class="fu">apply</span><span class="op">(</span><span class="va">counts_data</span> <span class="op">&gt;</span> <span class="fl">100</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Extract species names from taxonomic strings</span></span>
<span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"s__"</span>, <span class="st">""</span>, <span class="fu">sapply</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="fu">strsplit</span><span class="op">(</span><span class="va">y</span>, <span class="st">"\\|"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">counts_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">species_names</span></span>
<span></span>
<span><span class="co"># Create a taxonomic lineage data frame to define Sets</span></span>
<span><span class="co"># Remove taxonomic prefixes (k__, p__, c__, etc.) and separate into taxonomic ranks</span></span>
<span><span class="va">taxon_lineages</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span>Name <span class="op">=</span> <span class="va">species_names</span>,</span>
<span>                             Lineage <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>Lineage <span class="op">=</span> <span class="fu">str_remove_all</span><span class="op">(</span><span class="va">Lineage</span>, <span class="st">'[kpcofgs]__'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">separate</span><span class="op">(</span>col <span class="op">=</span> <span class="va">Lineage</span>, into <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">'kingdom'</span>, <span class="st">'phylum'</span>, <span class="st">'class'</span>, </span>
<span>                                   <span class="st">'order'</span>, <span class="st">'family'</span>, <span class="st">'genus'</span>, <span class="st">'species'</span><span class="op">)</span>, </span>
<span>           sep <span class="op">=</span> <span class="st">'\\|'</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>name <span class="op">=</span> <span class="va">Name</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">remove_rownames</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">column_to_rownames</span><span class="op">(</span><span class="st">'name'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform differential abundance testing using LinDA</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">study_condition</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">study_condition</span>, levels <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"AD"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">linda_results</span> <span class="op">&lt;-</span> <span class="fu">linda</span><span class="op">(</span></span>
<span>  feature.dat <span class="op">=</span> <span class="va">counts_data</span>,</span>
<span>  meta.dat <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>  formula <span class="op">=</span> <span class="st">'~study_condition'</span>,</span>
<span>  feature.dat.type <span class="op">=</span> <span class="st">'count'</span>,</span>
<span>  prev.filter <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>216  features are filtered!
The filtered data has  78  samples and  216  features will be tested!
Pseudo-count approach is used.
Fit linear models ...
Completed.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract log2 fold change values for differential taxa</span></span>
<span><span class="va">linda_results</span> <span class="op">&lt;-</span> <span class="va">linda_results</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">study_conditionAD</span></span>
<span><span class="va">log2_fold_changes</span> <span class="op">&lt;-</span> <span class="va">linda_results</span><span class="op">$</span><span class="va">log2FoldChange</span></span>
<span><span class="fu">names</span><span class="op">(</span><span class="va">log2_fold_changes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">linda_results</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Create a named list of species grouped by order</span></span>
<span><span class="va">custom_taxon_sets</span> <span class="op">&lt;-</span> <span class="va">taxon_lineages</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">.data</span><span class="op">[[</span><span class="st">'order'</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">summarise</span><span class="op">(</span>species <span class="op">=</span> <span class="fu">list</span><span class="op">(</span><span class="va">species</span><span class="op">)</span>, .groups <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">deframe</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform enrichment analysis using TaxSEA</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="fu">TaxSEA</span><span class="op">(</span>taxon_ranks <span class="op">=</span> <span class="va">log2_fold_changes</span>, custom_db <span class="op">=</span> <span class="va">custom_taxon_sets</span><span class="op">)</span></span>
<span><span class="va">custom_taxsea_results</span> <span class="op">&lt;-</span> <span class="va">custom_taxsea_results</span><span class="op">$</span><span class="va">custom_sets</span></span>
<span></span>
<span><span class="fu">datatable</span><span class="op">(</span><span class="va">custom_taxsea_results</span><span class="op">)</span></span></code></pre>
</div>
<!--html_preserve-->
<div id="htmlwidget-cccef5c8f0794023dc74" class="datatables html-widget html-fill-item" style="width:100%;height:auto;">

</div>
<script type="application/json" data-for="htmlwidget-cccef5c8f0794023dc74">{"x":{"filter":"none","vertical":false,"data":[["Pasteurellales","Corynebacteriales","Lactobacillales","Bacillales","Micrococcales","Rhodobacterales","Bacteroidales","Actinomycetales","Rhodospirillales","Pseudomonadales","Neisseriales","Propionibacteriales","Flavobacteriales"],["Pasteurellales","Corynebacteriales","Lactobacillales","Bacillales","Micrococcales","Rhodobacterales","Bacteroidales","Actinomycetales","Rhodospirillales","Pseudomonadales","Neisseriales","Propionibacteriales","Flavobacteriales"],[1.119543769360394,-0.3879824092840792,1.338184015953424,1.849569432709268,-0.6707002477465542,-1.411169277303825,0.6960170764832077,1.340148508123332,0.1321514169238954,-0.1249914266662796,0.6086800199417697,-0.3124833654946449,0.4187268857840517],[0.0118725802386351,0.01329696522186961,0.0260205602170516,0.04396131142078909,0.1293477081093432,0.1390625964594388,0.1580926108976503,0.1721450259759667,0.4374982180848367,0.4427549556643566,0.6065049269116362,0.8719580780499743,0.9521634436974848],[0.6171428571428571,0.2673170731707317,0.3218181818181818,0.3453571428571429,0.2542857142857143,0.4495238095238095,0.3,0.3428571428571429,0.3657142857142858,0.2285714285714285,0.2742857142857143,0.2457142857142857,0.1828571428571429],[0.08643027394215246,0.08643027394215246,0.1127557609405569,0.1428742621175645,0.2797356672109459,0.2797356672109459,0.2797356672109459,0.2797356672109459,0.5755814423636636,0.5755814423636636,0.7167785499864792,0.9446212512208054,0.9521634436974848]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>taxonSetName<\/th>\n      <th>median_rank_of_set_members<\/th>\n      <th>PValue<\/th>\n      <th>Test_statistic<\/th>\n      <th>FDR<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,5]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"taxonSetName","targets":1},{"name":"median_rank_of_set_members","targets":2},{"name":"PValue","targets":3},{"name":"Test_statistic","targets":4},{"name":"FDR","targets":5}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><!--/html_preserve--><div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plotting</span></span>
<span><span class="fu">data.frame</span><span class="op">(</span>Name <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="va">log2_fold_changes</span><span class="op">)</span>,</span>
<span>           Log2FC <span class="op">=</span> <span class="va">log2_fold_changes</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">taxon_lineages</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">custom_taxsea_results</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"order"</span> <span class="op">=</span> <span class="st">"taxonSetName"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>Direction <span class="op">=</span> <span class="fu">if_else</span><span class="op">(</span><span class="va">FDR</span> <span class="op">&gt;</span> <span class="fl">0.1</span>, <span class="st">'Insig.'</span>, <span class="fu">if_else</span><span class="op">(</span><span class="va">median_rank_of_set_members</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="st">'Up'</span>, <span class="st">'Down'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">drop_na</span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Log2FC</span>, y <span class="op">=</span> <span class="va">order</span>, group <span class="op">=</span> <span class="va">order</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_density_ridges</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span> <span class="va">Direction</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">'Insig.'</span> <span class="op">=</span> <span class="st">'grey80'</span>,</span>
<span>                              <span class="st">'Up'</span> <span class="op">=</span> <span class="st">'steelblue'</span>,</span>
<span>                              <span class="st">'Down'</span> <span class="op">=</span> <span class="st">'firebrick'</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/02-taxsea-rendered-unnamed-chunk-11-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul><li>TaxSEA performs taxon-set enrichment analysis on differential
abundance results, similar to GSEA.</li>
<li>Inputs must be species or genus names with corresponding ranks
(e.g., log2 fold changes).</li>
<li>TaxSEA tests whether members of a taxon set are skewed toward one
end of the ranked distribution.</li>
<li>Enrichment results help reveal functional or ecological patterns
that may not be visible at the single-taxon level.</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/01-metagenomic-workflows.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/03-multi-omics.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/01-metagenomic-workflows.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to
        </a>
        <a class="chapter-link float-end" href="../instructor/03-multi-omics.html" rel="next">
          Next: Multi-omics...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/edit/main/episodes/02-taxsea.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/" class="external-link">Source</a></p>
        <p>
        <a href="../citation.html">Cite</a>
        | <a href="mailto:vinicius.salazar%20%5B%20at%20%5D%20unimelb.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors.</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/melbournebioinformatics/uom-varnish/tree/main" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "melbournebioinformatics.github.io/uom-varnish/instructor/02-taxsea.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "metagenomics, microbiome, bioinformatics, multiomics, metabolomics, enrichment analysis",
  "name": "Taxon-set enrichment analysis with TaxSEA",
  "creativeWorkStatus": "active",
  "url": "melbournebioinformatics.github.io/uom-varnish/instructor/02-taxsea.html",
  "identifier": "melbournebioinformatics.github.io/uom-varnish/instructor/02-taxsea.html",
  "dateCreated": "2025-10-16",
  "dateModified": "2025-11-25",
  "datePublished": "2025-11-27"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

